debug/debug.go

```diff
@@ -0,0 +1,296 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	expr "github.com/google/cel-spec/proto/v1/syntax"
+	"strconv"
+)
+
+// DebugAdorner returns debug metadata that will be tacked on to the string
+// representation of an expression.
+type DebugAdorner interface {
+	GetMetadata(e interface{}) string
+}
+
+type DebugWriter interface {
+	fmt.Stringer
+	Append(e *expr.Expr)
+}
+
+type emptyDebugAdorner struct {
+}
+
+var EmptyAdorner DebugAdorner = &emptyDebugAdorner{}
+
+func (a *emptyDebugAdorner) GetMetadata(e interface{}) string {
+	return ""
+}
+
+func ToDebugString(e *expr.Expr) string {
+	return ToAdornedDebugString(e, EmptyAdorner)
+}
+
+func ToAdornedDebugString(e *expr.Expr, adorner DebugAdorner) string {
+	w := newDebugWriter(adorner)
+	w.Append(e)
+	return w.String()
+}
+
+// debugWriter is used to print out pretty-printed debug strings.
+type debugWriter struct {
+	adorner   DebugAdorner
+	buffer    bytes.Buffer
+	indent    int
+	lineStart bool
+}
+
+func newDebugWriter(a DebugAdorner) *debugWriter {
+	return &debugWriter{
+		adorner:   a,
+		indent:    0,
+		lineStart: true,
+	}
+}
+
+func (w *debugWriter) Append(e *expr.Expr) {
```

Comment:
Not really an idiom, so much as that's how Go does visibility (Caps is public, lower is package private). I can rename the method to something else if it would make things clearer.

Reply:
Yeah, unless it's a well-established idiom, I'd prefer a rename.

Reply:
Renamed to Buffer()